spec/abci++/abci++_properties_001_draft.md

```diff
@@ -0,0 +1,675 @@
+
+### ProcessProposal
+
+#### Parameters and Types
+
+* **Request**:
+
+    | Name                 | Type                                        | Description                                                                                                 | Field Number |
+    |----------------------|---------------------------------------------|-------------------------------------------------------------------------------------------------------------|--------------|
+    | hash                 | bytes                                       | The proposed block's hash. This can be derived from the block header.                                       | 1            |
+    | header               | [Header](../core/data_structures.md#header) | The proposed block's header.                                                                                | 2            |
+    | last_commit_info     | [LastCommitInfo](#lastcommitinfo)           | Info about the last commit, including the round , the validator list, and which ones signed the last block. | 3            |
+    | byzantine_validators | repeated [Evidence](#evidence)              | List of evidence of validators that acted maliciously.                                                      | 4            |
+    | tx                   | repeated bytes                              | List of transactions that have been picked as part of the proposed block.                                   | 5            |
+
+* **Response**:
+
+    | Name   | Type | Description                                                      | Field Number |
+    |--------|------|------------------------------------------------------------------|--------------|
+    | accept | bool | If false, instruct Tendermint to prevote $\bot$ for this proposal | 1            |
+
+* **Usage**:
+    * Contains a full proposed block.
+        * The parameters and types of `RequestProcessProposal` are the same as `RequestFinalizeBlock`
+          (**TODO**: We'll see...).
+        * In "same-block execution" mode, the Application will fully execute the block
+          as though it was handling `RequestFinalizeBlock`.
+          However, any resulting state changes must be kept as _canditade state_,
+          and the Application should be ready to
+          backtrack/discard it in case the decided block is different.
+    * If `ResponseProcessProposal.accept` is true, Tendermint should prevote according to
+      the normal procedure; else, it should prevote $\bot$.
+    * The [Header](../core/data_structures.md#header) contains a boolan `problem` that indicates
+      if the block is the result of an Application rejection (i.e., $\bot$). This allows the
+      users to distinguish empty blocks because of lack of activity and empty blocks representing
+      $\bot$, i.e., a problem in the code of `PrepareProposal` and/or `ProcessProposal`.
+    * The implementation of `ProcessProposal` MUST be deterministic. Moreover, the value of
```

Comment:
Should we bypass this logic if `validValue != nil`? Perhaps an mild optimization if this method must be deterministic.